File: Coding_ObjectiveC/CircularLinkedList.m
//
//  CircularLinkedList.m
//  Coding_ObjectiveC
//
//  Created by Alexander Kobylinskyy on 1/2/14.
//  Copyright (c) 2014 Alex. All rights reserved.
//

#import "CircularLinkedList.h"
#import "Node.h"

@implementation CircularLinkedList

@synthesize head = _head;
@synthesize tail = _tail;

#pragma mark - ListProtocol

- (void)insertNode:(Node *)node {
	if (!_head) {
		_head = node;
		_tail = _head;
	}
	else {
		_tail.next = node;
		_tail = _tail.next;
	}
	_tail.next = _head;
}

- (Node *)findNodeWithData:(NSInteger)data {
	Node *resultNode = nil;
	for (Node *node = _head; node != nil; node = node.next) {
		if (node.data == data) {
			resultNode = node;
			break;
		}
		if (node.next == _head) {
			break;
		}
	}
	return resultNode;
}

- (Node *)findBeforeNodeWithData:(NSInteger)data {
	Node *resultNode = nil;
	for (Node *node = _head; node != nil; node = node.next) {
		if (node.next.data == data) {
			resultNode = node;
			break;
		}
		if (node.next == _head) {
			break;
		}
	}
	return resultNode;
}

- (void)removeNodeWithData:(NSInteger)data {
	if (data == _head.data) {
		Node *resultNode = _head;
		if (_head.next == _head) {
			_head = nil;
		}
		else
		{
			_head =  _head.next;
			_tail.next = _head;
		}
		if (resultNode == _tail) {
			_tail = nil;
		}
		resultNode = nil;
		return;
	}
	Node *node = [self findBeforeNodeWithData:data];
	if (node) {
		Node *resultNode = node.next;
		node.next = node.next.next;
		if (_tail == resultNode) {
			_tail = node;
		}
		resultNode = nil;
	}
}

#pragma mark - Description

- (NSString *)description {
	NSMutableString *description = [NSMutableString string];
	Node *node = _head;
    while (node) {
		[description appendFormat:@"%@ ", node.description];
        node = node.next;
		if (node == _head) {
			break;
		}
    }
	return [description copy];
}

@end

File: SWDataStructure/CircularLinkedList.swift
//
//  CircularLinkedList.swift
//  Coding_ObjectiveC
//
//  Created by Alexander Kobylinskyy on 9/3/14.
//  Copyright (c) 2014 Alex. All rights reserved.
//

class CircularLinkedList<T: Equatable>: LinkedList<T> {
	
	override var tail: Node<T>? {
		get {
			var node = head
			while let _ = node?.next {
				if (node?.next == head) {
					break
				}
				node = node?.next
			}
			return node
		}
	}
	
	override init(head: Node<T>) {
		super.init(head: head)
	}
	
	override var count: Int {
		var count = 0
		for var node: Node<T>? = head; node != nil; node = node?.next {
			count++
			if node?.next == head {
				break
			}
		}
		return count
	}
	
	override func insertNode(node: Node<T>) {
		let tailNode = self.tail
		if let _ = tailNode {
			tailNode!.next = node
			node.next = head
		}
		else {
			head = node
			head?.next = head
		}
	}
	
	override func removeNode(node: Node<T>) {
		var prevNode: Node<T>? = nil
		let tailNode = self.tail
		for var listNode: Node<T>? = head; listNode != nil; prevNode = listNode, listNode = listNode?.next {
			if node == listNode {
				if (node == head) {
					if (head?.next == head) {
						head = nil
					}
					else {
						head = head?.next
						tailNode?.next = head
					}
				}
				else {
					prevNode?.next = prevNode?.next?.next
				}
				listNode = nil
				break
			}
		}
	}
	
	override func findData(data: T) -> Node<T>? {
		var result: Node<T>? = nil
		for var node: Node<T>? = head; node != nil; node = node?.next {
			if node?.data == data {
				result = node
				break
			}
			if node?.next == head {
				break
			}
		}
		return result
	}
	
	override func reverse() {
	}
	
	override var description: String {
		var result: String = ""
		for var node: Node<T>? = head; node != nil; node = node?.next {
			result += "\(node)"
			result += " "
			if node?.next == head {
				break
			}
		}
		return result
	}
	
}
